File: sage-boi-service/SageBOI.Api/Controllers/HealthController.cs
using Microsoft.AspNetCore.Mvc;
using SageBOI.Api.Services;

namespace SageBOI.Api.Controllers;

[ApiController]
[Route("[controller]")]
public class HealthController : ControllerBase
{
    private readonly IProvideXSessionManager _sessionManager;
    private readonly ILogger<HealthController> _logger;

    public HealthController(
        IProvideXSessionManager sessionManager,
        ILogger<HealthController> logger)
    {
        _sessionManager = sessionManager;
        _logger = logger;
    }

    /// <summary>
    /// Basic health check
    /// </summary>
    [HttpGet]
    public IActionResult Get()
    {
        return Ok(new
        {
            status = "healthy",
            service = "sage-boi-service",
            timestamp = DateTime.UtcNow
        });
    }

    /// <summary>
    /// Detailed health check including Sage 100 connectivity
    /// </summary>
    [HttpGet("ready")]
    public async Task<IActionResult> Ready()
    {
        try
        {
            var isHealthy = await _sessionManager.IsHealthyAsync();
            
            if (isHealthy)
            {
                return Ok(new
                {
                    status = "ready",
                    sage100 = "connected",
                    timestamp = DateTime.UtcNow
                });
            }
            else
            {
                return StatusCode(503, new
                {
                    status = "not_ready",
                    sage100 = "unavailable",
                    timestamp = DateTime.UtcNow
                });
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Health check failed");
            return StatusCode(503, new
            {
                status = "not_ready",
                error = ex.Message,
                timestamp = DateTime.UtcNow
            });
        }
    }
}

File: api/src/controllers/health.controller.js
import queueService from '../services/queue.service.js';
import boiClient from '../services/boiClient.service.js';
import logger from '../utils/logger.js';

/**
 * Basic health check
 */
export const healthCheck = (req, res) => {
  res.json({
    status: 'healthy',
    service: 'uaf-sage-api',
    timestamp: new Date().toISOString(),
  });
};

/**
 * Readiness check - includes dependencies
 */
export const readinessCheck = async (req, res) => {
  const checks = {
    redis: false,
    boiService: false,
  };

  try {
    // Check Redis
    checks.redis = queueService.isConnected();

    // Check BOI Service
    checks.boiService = await boiClient.isHealthy();

    // Get queue depth
    const queueDepth = await queueService.getQueueDepth();

    const allHealthy = Object.values(checks).every(check => check === true);

    const response = {
      status: allHealthy ? 'ready' : 'not_ready',
      checks,
      queueDepth,
      timestamp: new Date().toISOString(),
    };

    if (allHealthy) {
      res.json(response);
    } else {
      res.status(503).json(response);
    }
  } catch (error) {
    logger.error('Readiness check failed', { error: error.message });
    
    res.status(503).json({
      status: 'not_ready',
      checks,
      error: error.message,
      timestamp: new Date().toISOString(),
    });
  }
};

export default {
  healthCheck,
  readinessCheck,
};

